### PR TITLE
ENH: dump runtime stats to cloud storage systems

### LINKED ISSUE #214: ENH: dump collected runtime stats to cloud storage
### Is your feature request related to a problem? Please describe
Store runtime stats to cloud storage.

### PR DESCRIPTION
**Pull request by UranusSeven**
## Related issue number

Resolves #214 

## Check code requirements

- [ ] tests added / passed (if needed)
- [ ] Ensure all linting tests pass

### CODE REVIEW

File: python/xorbits/_mars/services/task/supervisor/tests/test_task_manager.py [hunk 1]
```diff
@@ -115,6 +118,7 @@ async def actor_pool():
             await MockStorageAPI.cleanup(pool.external_address)
             await MockClusterAPI.cleanup(pool.external_address)
             await MockMutableAPI.cleanup(session_id, pool.external_address)
+            shutil.rmtree(mars_tmp_dir, ignore_errors=True)
```
Comment (aresnow1) [inline python/xorbits/_mars/services/task/supervisor/tests/test_task_manager.py:121]: Will `mars_tmp_dir` not be deleted during cleanup?

File: python/xorbits/_mars/services/task/supervisor/tests/test_task_manager.py [hunk 2]
```diff
@@ -738,44 +742,16 @@ async def test_collect_task_info(actor_pool):
...
-        assert "fetch_time" in v
-
-    shutil.rmtree(save_dir, ignore_errors=True)
+    time.sleep(1)
+
+    files = os.listdir(save_dir)
+    assert "tileables.yaml" in files
+    assert "result_nodes.yaml" in files
+    assert len([f for f in files if f.endswith("_subtask_info.yaml")]) > 0
+    assert len([f for f in files if f.endswith("_subtask_fetch_time.yaml")]) > 0
```
Comment (aresnow1) [inline python/xorbits/_mars/services/task/supervisor/tests/test_task_manager.py:757]: Do we no longer need those deleted checks?

File: python/xorbits/_mars/services/task/task_info_collector.py [hunk 3]
```diff
@@ -139,46 +150,48 @@ async def collect_subtask_operand_structure(
...
-            subtask_save_path = os.path.join(
-                self._save_dir, f"{stage_id[:trunc_key]}_subtask.yaml"
-            )
-            await self._save_task_info(
-                {subtask.subtask_id: subtask_dict}, subtask_save_path
-            )
+            subtask_info[subtask.subtask_id] = subtask_info
+
+        await self._save_task_info(op_info, op_save_path)
+        await self._save_task_info(subtask_info, subtask_save_path)
```
Comment (aresnow1) [inline python/xorbits/_mars/services/task/task_info_collector.py:161]: Do we need to await this call? Or create an async task in `self._save_task_info`.

File: python/xorbits/_mars/services/task/task_info_collector.py [hunk 4]
```diff
@@ -299,36 +328,50 @@ class TaskInfoCollectorActor(mo.Actor):
...
+        self._task_info_root_path = experimental_profiling_config.get(
+            "task_info_root_path", None
+        )
+        self._task_info_storage_options = experimental_profiling_config.get(
+            "task_info_storage_options", {}
+        )
+        if self._task_info_root_path is None:
+            profiling_results_dir = os.environ.get(MARS_PROFILING_RESULTS_PATH_KEY)
+            if profiling_results_dir is None:
+                profiling_results_dir = tempfile.mkdtemp(prefix=MARS_TMP_DIR_PREFIX)
```
Comment (aresnow1) [inline python/xorbits/_mars/services/task/task_info_collector.py:345]: Do we need to manage these temporary paths? log and profiling use different paths.

**Comment by aresnow1** [comment]
So we dump each subtask info into one file now? Is there a significant overhead?

File: python/xorbits/_mars/services/task/task_info_collector.py (hunk 4 above)
Comment (UranusSeven) [reply python/xorbits/_mars/services/task/task_info_collector.py:345]: True. How about /tmp/mars_profiling ?

File: python/xorbits/_mars/services/task/supervisor/tests/test_task_manager.py (hunk 1 above)
Comment (UranusSeven) [reply python/xorbits/_mars/services/task/supervisor/tests/test_task_manager.py:121]: The temp dir was removed in the end of `test_collect_task_info`, which is not reasonable. If the UT failed, the temp dir will not be removed.

File: python/xorbits/_mars/services/task/task_info_collector.py (hunk 4 above)
Comment (aresnow1) [reply python/xorbits/_mars/services/task/task_info_collector.py:345]: I'm thinking to put them under the same directory:
- /tmp/mars
   - mars_log
   - mars_profiling

Comment (UranusSeven) [reply python/xorbits/_mars/services/task/task_info_collector.py:345]: After discussion, we agreed that the temp dir root path should be /tmp/mars on macOS / Linux. The logs should be in /tmp/mars/log and profiling results should be in /tmp/mars/profiling

File: python/xorbits/_mars/services/task/task_info_collector.py [hunk 5]
```diff
@@ -139,46 +150,50 @@ async def collect_subtask_operand_structure(
...
-            subtask_save_path = os.path.join(
-                self._save_dir, f"{stage_id[:trunc_key]}_subtask.yaml"
-            )
-            await self._save_task_info(
-                {subtask.subtask_id: subtask_dict}, subtask_save_path
-            )
+            subtask_info[subtask.subtask_id] = subtask_info
+
+        asyncio.create_task(self._save_task_info(op_info, op_save_path))
+        asyncio.create_task(self._save_task_info(subtask_info, subtask_save_path))
```
Comment (aresnow1) [inline python/xorbits/_mars/services/task/task_info_collector.py:161]: As all collections call `_save_task_info`, just modify the `_save_task_info` function.
`await task_api.save_task_info(task_info, path)` ->
`asyncio.create_task(task_api.save_task_info(task_info, path))`

File: python/xorbits/_mars/services/task/task_info_collector.py [hunk 6]
```diff
@@ -139,46 +150,50 @@ async def collect_subtask_operand_structure(
...
-        save_path = os.path.join(self._save_dir, "last_nodes.yaml")
-        await self._save_task_info(
-            {"op": last_op_keys, "subtask": last_subtask_keys},
-            save_path,
+                result_subtask_keys.append(subtask.subtask_id)
+                for result in subtask.chunk_graph.results:
+                    result_op_keys.append(result.op.key)
+
+        save_path = os.path.join(self._save_dir, "result_nodes.yaml")
+        asyncio.create_task(
```
Comment (aresnow1) [inline python/xorbits/_mars/services/task/task_info_collector.py:183]: Ditto.

File: python/xorbits/_mars/services/task/task_info_collector.py [hunk 7]
```diff
@@ -189,11 +204,11 @@ async def collect_tileable_structure(
                 x.key for x in tileable.op.inputs
             ]
 
-        save_path = os.path.join(self._save_dir, "tileable.yaml")
-        await self._save_task_info(tileable_dict, save_path)
+        save_path = os.path.join(self._save_dir, "tileables.yaml")
+        asyncio.create_task(self._save_task_info(tileable_dict, save_path))
```
Comment (aresnow1) [inline python/xorbits/_mars/services/task/task_info_collector.py:208]: Ditto.

File: python/xorbits/_mars/services/task/task_info_collector.py [hunk 8]
```diff
@@ -255,23 +275,36 @@ async def collect_runtime_operand_info(
...
+            self._subtask_infos[subtask.subtask_id]["op_info"] = dict()
+        self._subtask_infos[subtask.subtask_id]["op_info"][op_key] = op_info
+
+    @collect_on_demand
+    async def flush_subtask_info(self, subtask_id: str):
+        if subtask_id in self._subtask_infos:
+            save_path = os.path.join(self._save_dir, f"{subtask_id}_subtask_info.yaml")
+            subtask_info = self._subtask_infos[subtask_id]
+            del self._subtask_infos[subtask_id]
+            asyncio.create_task(self._save_task_info(subtask_info, save_path))
```
Comment (aresnow1) [inline python/xorbits/_mars/services/task/task_info_collector.py:288]: Ditto.

File: python/xorbits/_mars/services/task/task_info_collector.py [hunk 9]
```diff
@@ -255,23 +275,36 @@ async def collect_runtime_operand_info(
...
-            {
-                subtask.subtask_id: {
-                    "fetch_time": {"start_time": fetch_start, "end_time": fetch_end}
-                }
-            },
-            save_path,
+        save_path = os.path.join(
+            self._save_dir, f"{subtask.subtask_id}_subtask_fetch_time.yaml"
+        )
+        asyncio.create_task(
```
Comment (aresnow1) [inline python/xorbits/_mars/services/task/task_info_collector.py:303]: Ditto.

**Comment by UranusSeven** [comment]
> So we dump each subtask info into one file now? Is there a significant overhead?

Right. There'll be a file for each subtask. The overhead is about 7% when running on my laptop with TPC-H SF10.

**Review by aresnow1 — DISMISSED** [review]
LGTM

**Review by aresnow1 — APPROVED** [review]
LGTM

**Review by ChengjieLi28 — APPROVED** [review]

**Pull request merged**